File: src/main/java/com/refinedmods/refinedstorage/tile/data/RSSerializers.java
package com.refinedmods.refinedstorage.tile.data;

import com.refinedmods.refinedstorage.api.storage.AccessType;
import com.refinedmods.refinedstorage.tile.ClientNode;
import com.refinedmods.refinedstorage.util.AccessTypeUtils;
import net.minecraft.network.PacketBuffer;
import net.minecraft.network.datasync.DataParameter;
import net.minecraft.network.datasync.IDataSerializer;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fluids.FluidStack;

import java.util.*;

public final class RSSerializers {
    private RSSerializers() {
    }

    public static final IDataSerializer<List<ClientNode>> CLIENT_NODE_SERIALIZER = new IDataSerializer<List<ClientNode>>() {
        @Override
        public void write(PacketBuffer buf, List<ClientNode> nodes) {
            buf.writeInt(nodes.size());

            for (ClientNode node : nodes) {
                buf.writeItem(node.getStack());
                buf.writeInt(node.getAmount());
                buf.writeInt(node.getEnergyUsage());
            }
        }

        @Override
        public List<ClientNode> read(PacketBuffer buf) {
            List<ClientNode> nodes = new ArrayList<>();

            int size = buf.readInt();

            for (int i = 0; i < size; ++i) {
                nodes.add(new ClientNode(buf.readItem(), buf.readInt(), buf.readInt()));
            }

            return nodes;
        }

        @Override
        public DataParameter<List<ClientNode>> createAccessor(int id) {
            return null;
        }

        @Override
        public List<ClientNode> copy(List<ClientNode> value) {
            return value;
        }
    };

    public static final IDataSerializer<FluidStack> FLUID_STACK_SERIALIZER = new IDataSerializer<FluidStack>() {
        @Override
        public void write(PacketBuffer buf, FluidStack value) {
            value.writeToPacket(buf);
        }

        @Override
        public FluidStack read(PacketBuffer buf) {
            return FluidStack.readFromPacket(buf);
        }

        @Override
        public DataParameter<FluidStack> createAccessor(int id) {
            return null;
        }

        @Override
        public FluidStack copy(FluidStack value) {
            return value;
        }
    };

    public static final IDataSerializer<AccessType> ACCESS_TYPE_SERIALIZER = new IDataSerializer<AccessType>() {
        @Override
        public void write(PacketBuffer buf, AccessType value) {
            buf.writeInt(value.getId());
        }

        @Override
        public AccessType read(PacketBuffer buf) {
            return AccessTypeUtils.getAccessType(buf.readInt());
        }

        @Override
        public DataParameter<AccessType> createAccessor(int id) {
            return null;
        }

        @Override
        public AccessType copy(AccessType value) {
            return value;
        }
    };

    public static final IDataSerializer<Long> LONG_SERIALIZER = new IDataSerializer<Long>() {
        @Override
        public void write(PacketBuffer buf, Long value) {
            buf.writeLong(value);
        }

        @Override
        public Long read(PacketBuffer buf) {
            return buf.readLong();
        }

        @Override
        public DataParameter<Long> createAccessor(int id) {
            return null;
        }

        @Override
        public Long copy(Long value) {
            return value;
        }
    };

    public static final IDataSerializer<Optional<ResourceLocation>> OPTIONAL_RESOURCE_LOCATION_SERIALIZER = new IDataSerializer<Optional<ResourceLocation>>() {
        @Override
        public void write(PacketBuffer buf, Optional<ResourceLocation> value) {
            buf.writeBoolean(value.isPresent());

            value.ifPresent(buf::writeResourceLocation);
        }

        @Override
        public Optional<ResourceLocation> read(PacketBuffer buf) {
            if (!buf.readBoolean()) {
                return Optional.empty();
            }

            return Optional.of(buf.readResourceLocation());
        }

        @Override
        public DataParameter<Optional<ResourceLocation>> createAccessor(int id) {
            return null;
        }

        @Override
        public Optional<ResourceLocation> copy(Optional<ResourceLocation> value) {
            return value;
        }
    };

    public static final IDataSerializer<List<Set<ResourceLocation>>> LIST_OF_SET_SERIALIZER = new IDataSerializer<List<Set<ResourceLocation>>>() {
        @Override
        public void write(PacketBuffer buf, List<Set<ResourceLocation>> value) {
            buf.writeInt(value.size());

            for (Set<ResourceLocation> values : value) {
                buf.writeInt(values.size());

                values.forEach(buf::writeResourceLocation);
            }
        }

        @Override
        public List<Set<ResourceLocation>> read(PacketBuffer buf) {
            List<Set<ResourceLocation>> value = new ArrayList<>();

            int size = buf.readInt();
            for (int i = 0; i < size; ++i) {
                int setSize = buf.readInt();

                Set<ResourceLocation> values = new HashSet<>();

                for (int j = 0; j < setSize; ++j) {
                    values.add(buf.readResourceLocation());
                }

                value.add(values);
            }

            return value;
        }

        @Override
        public DataParameter<List<Set<ResourceLocation>>> createAccessor(int id) {
            return null;
        }

        @Override
        public List<Set<ResourceLocation>> copy(List<Set<ResourceLocation>> value) {
            return value;
        }
    };
}

File: src/main/kotlin/com/refinedmods/refinedstorage/tile/data/RSSerializers.kt
package com.refinedmods.refinedstorage.tile.data

import com.refinedmods.refinedstorage.api.storage.AccessType
import com.refinedmods.refinedstorage.tile.ClientNode
import com.refinedmods.refinedstorage.util.AccessTypeUtils
import net.minecraft.network.PacketByteBuf
import net.minecraft.network.datasync.DataParameter
import net.minecraft.network.datasync.PacketIO
import net.minecraft.util.Identifier
import net.minecraftforge.fluids.FluidInstance
import java.util.*

object RSSerializers {
    val CLIENT_NODE_SERIALIZER: PacketIO<List<ClientNode>> = object : PacketIO<List<ClientNode?>?>() {
        fun write(buf: PacketByteBuf, nodes: List<ClientNode>) {
            buf.writeInt(nodes.size)
            for (node in nodes) {
                buf.writeItemStack(node.stack)
                buf.writeInt(node.amount)
                buf.writeInt(node.energyUsage)
            }
        }

        fun read(buf: PacketByteBuf): List<ClientNode> {
            val nodes: MutableList<ClientNode> = ArrayList()
            val size: Int = buf.readInt()
            for (i in 0 until size) {
                nodes.add(ClientNode(buf.readItemStack(), buf.readInt(), buf.readInt()))
            }
            return nodes
        }

        fun createKey(id: Int): DataParameter<List<ClientNode>>? {
            return null
        }

        fun copyValue(value: List<ClientNode>): List<ClientNode> {
            return value
        }
    }
    val FLUID_STACK_SERIALIZER: PacketIO<FluidInstance> = object : PacketIO<FluidInstance?>() {
        fun write(buf: PacketByteBuf?, value: FluidInstance) {
            value.writeToPacket(buf)
        }

        fun read(buf: PacketByteBuf?): FluidInstance {
            return FluidInstance.readFromPacket(buf)
        }

        fun createKey(id: Int): DataParameter<FluidInstance>? {
            return null
        }

        fun copyValue(value: FluidInstance): FluidInstance {
            return value
        }
    }
    val ACCESS_TYPE_SERIALIZER: PacketIO<AccessType> = object : PacketIO<AccessType?>() {
        fun write(buf: PacketByteBuf, value: AccessType) {
            buf.writeInt(value.getId())
        }

        fun read(buf: PacketByteBuf): AccessType {
            return AccessTypeUtils.getAccessType(buf.readInt())
        }

        fun createKey(id: Int): DataParameter<AccessType>? {
            return null
        }

        fun copyValue(value: AccessType): AccessType {
            return value
        }
    }
    val LONG_SERIALIZER: PacketIO<Long> = object : PacketIO<Long?>() {
        fun write(buf: PacketByteBuf, value: Long?) {
            buf.writeLong(value)
        }

        fun read(buf: PacketByteBuf): Long {
            return buf.readLong()
        }

        fun createKey(id: Int): DataParameter<Long>? {
            return null
        }

        fun copyValue(value: Long): Long {
            return value
        }
    }
    val OPTIONAL_RESOURCE_LOCATION_SERIALIZER: PacketIO<Optional<Identifier>> = object : PacketIO<Optional<Identifier?>?>() {
        fun write(buf: PacketByteBuf, value: Optional<Identifier?>) {
            buf.writeBoolean(value.isPresent())
            value.ifPresent(buf::writeIdentifier)
        }

        fun read(buf: PacketByteBuf): Optional<Identifier> {
            return if (!buf.readBoolean()) {
                Optional.empty<Identifier>()
            } else Optional.of(buf.readIdentifier())
        }

        fun createKey(id: Int): DataParameter<Optional<Identifier>>? {
            return null
        }

        fun copyValue(value: Optional<Identifier>): Optional<Identifier> {
            return value
        }
    }
    val LIST_OF_SET_SERIALIZER: PacketIO<List<Set<Identifier>>> = object : PacketIO<List<Set<Identifier?>?>?>() {
        fun write(buf: PacketByteBuf, value: List<Set<Identifier?>>) {
            buf.writeInt(value.size)
            for (values in value) {
                buf.writeInt(values.size)
                values.forEach(buf::writeIdentifier)
            }
        }

        fun read(buf: PacketByteBuf): List<Set<Identifier>> {
            val value: MutableList<Set<Identifier>> = ArrayList<Set<Identifier>>()
            val size: Int = buf.readInt()
            for (i in 0 until size) {
                val setSize: Int = buf.readInt()
                val values: MutableSet<Identifier> = HashSet<Identifier>()
                for (j in 0 until setSize) {
                    values.add(buf.readIdentifier())
                }
                value.add(values)
            }
            return value
        }

        fun createKey(id: Int): DataParameter<List<Set<Identifier>>>? {
            return null
        }

        fun copyValue(value: List<Set<Identifier>>): List<Set<Identifier>> {
            return value
        }
    }
}